pyformex/extra/dxfparser/dxfparser.1.rst
==================
pyformex-dxfparser
==================

--------------------------------------------------------
Convert items from a .dxf file to easily readable format
--------------------------------------------------------

:Author: Benedict Verhegghe <benedict.verhegghe@ugent.be>. This manual page was written for the Debian project (and may be used by others).
:Date:   2012-08-08
:Copyright: GPL v3 or higher
:Version: 0.1
:Manual section: 1
:Manual group: text and X11 processing

SYNOPSIS
========

pyformex-dxfparser [OPTION] FILE ...

DESCRIPTION
===========

pyformex-dxfparser parses the DXF files specified on the command line
and prints out the recognized items in an easily readable and further
processable format.
Currently recognized items are: LINE, POLYLINE, ARC, CIRCLE.

OPTIONS
=======

--copyright      Print the copyright string and exit.
--version        Print the version string and exit.


AUTHOR
======

pyformex-dxfparser was written by Benedict Verhegghe <benedict.verhegghe@ugent.be> for the pyFormex project <https://savannah.nongnu.org/projects/pyformex/>.



History
# $Id$
##
##  This file is part of the pyFormex project.
##  pyFormex is a tool for generating, manipulating and transforming 3D
##  geometrical models by sequences of mathematical operations.
##  Home page: http://pyformex.org
##  Project page:  https://savannah.nongnu.org/projects/pyformex/
##  Copyright (C) Benedict Verhegghe (benedict.verhegghe@ugent.be) 
##  Distributed under the GNU General Public License version 3 or later.
##
##
##  This program is free software: you can redistribute it and/or modify
##  it under the terms of the GNU General Public License as published by
##  the Free Software Foundation, either version 3 of the License, or
##  (at your option) any later version.
##
##  This program is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License
##  along with this program.  If not, see http://www.gnu.org/licenses/.
##

2009-06 BV: Released pyFormex 0.8

2008-04 BV: Released pyFormex 0.7

2008-02 BV: Started the pyFormex User Meeting

2007-11 BV: Released pyFormex 0.6

2007-08 BV: Released pyFormex 0.5, a major step towards a multifunctional
            program.

2007-02 BV: Released pyFormex 0.4.2, including lots of work on surface
            models.

2006-11-03 BV: Released pyFormex 0.4. This release is marked by the transition
               to the Qt4 widget set.

2006-04-15 BV: Released pyFormex 0.3

2006: pyFormex was used in some of our research projects and proved to be
      a valuable tool. This necessitated and inspired new development work
      on pyFormex. Lots of new features were implemented, too many to be
      listed here.
      Some of the milestones:
      - pyFormex was used in the design of wire stents (see WireStent example).
        The process involved the automatic creation of Finite Element
        Analysis jobs and automatic processing of the results.
      - pyFormex was used to measure geometries obtained from medical images
        and to create a patient specific geometrical hardware model. This
        initiated our further work with surface models.
      - pyFormex was used by civil engineering students for their projects
        in designing shell structures. Some of these will be included as
        examples in future releases. 


2005-12-09 BV : First anniversary release 0.2.2
2005-04-23 BV : Interim 0.2.1 release with lots of new stuff
2005-01-03 BV : First public release on berlios.de (0.2)
2004-12-09 BV : Registered project pyFormex on berlios.de
2004-07-08 BV : Created a first release (0.1.1)
2004-07-02 BV : Created a Subversion repository
2004-06-29 BV :	After reading about Formian, I quickly hacked together
                the Formex class

# End

pyformex/extra/dxfparser/dxfparser.cc
/* $Id$ */
//
//  This file is part of pyFormex 0.8.5  (Sun Dec  4 21:24:46 CET 2011)
//  pyFormex is a tool for generating, manipulating and transforming 3D
//  geometrical models by sequences of mathematical operations.
//  Home page: http://pyformex.org
//  Project page:  http://savannah.nongnu.org/projects/pyformex/
//  Copyright 2004-2011 (C) Benedict Verhegghe (benedict.verhegghe@ugent.be) 
//  Distributed under the GNU General Public License version 3 or later.
//
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see http://www.gnu.org/licenses/.
//


/**************************************************************************
  dxfparser.cc      DXF parser
 
  dxfparser uses the dxflib from QCAD to parse an AutoCAD .DXF file,
  and exports it as a script with function calling syntax. 
 
***************************************************************************/

#include "dxfparser.h"

#include <stdio.h>

const char* _version_ = "dxfparser 0.2";
const char* _copyright_ = "Copyright (C) 2011 Benedict Verhegghe";
const char* ArcFmt = "Arc(%f,%f,%f,%f,%f,%f)\n";
const char* CircleFmt = "Circle(%f,%f,%f,%f)\n";
const char* LineFmt = "Line(%f,%f,%f,%f,%f,%f)\n";
const char* PolylineFmt = "Polyline(%d)\n";
const char* VertexFmt = "Vertex(%f,%f,%f)\n";

void MyDxfFilter::addArc(const DL_ArcData& d) {
  printf(ArcFmt,d.cx,d.cy,d.cz,d.radius,d.angle1,d.angle2);
}

void MyDxfFilter::addCircle(const DL_CircleData& d) {
  printf(CircleFmt,d.cx,d.cy,d.cz,d.radius);
}

void MyDxfFilter::addLine(const DL_LineData& d) {
  printf(LineFmt,d.x1,d.y1,d.z1,d.x2,d.y2,d.z2);
}

void MyDxfFilter::addPolyline(const DL_PolylineData& d){
  printf(PolylineFmt,d.number);
}

void MyDxfFilter::addVertex(const DL_VertexData& d){
  printf(VertexFmt,d.x,d.y,d.z);
}



int main(int argc, char* argv[])
{
  MyDxfFilter f;
  DL_Dxf dxf;
  for(int i = 1; i < argc; i++) {
    if (argv[i][0] == '-') {
      // Process switch
      if (!strcmp(argv[i],"--version")) {
	printf("%s\n",_version_);
	return 0;
      }
      if (!strcmp(argv[i],"--copyright")) {
	printf("%s\n",_copyright_);
	return 0;
      }
      printf("Unknown switch '%s'\n",argv[i]);
      return 1;
    }
    printf("# Converted from %s by %s\n",argv[i],_version_);
    if (!dxf.in(argv[i], &f)) {
      printf(" !! file could not be opened.\n");
      return 1;
    }
  }
  return 0;
}

// End
